In [1]:
from network import *
from framework import *
from preprocessing import *

cpu


In [2]:
isles_data_root = '/home/madtoinou/pdm/data/federated/synthetic/'
modality = 'Tmax'
batch_size = 2
#num_epochs = 300
#learning_rate = 0.000932#lrs[0] #To comment in the loop
#weighting_scheme = 'FEDAVG'
#beta_val=0.9

In [3]:
clients=["center1", "center2", "center4"]
#from SCAFFOLD manuscript, global_learning_rate should be = sqrt(#Samples sites)
local_epoch, global_epoch = 2, 20
#no sampling
K=len(clients)

local_lr, global_lr = 0.00932, 1.7 #np.sqrt(K)

_, centers_data_loaders, all_test_loader, _ = dataPreprocessing(isles_data_root, modality, 4, 2)

#move center 3 at the end of the dataloaders
tmp = centers_data_loaders[2]
centers_data_loaders[2]=centers_data_loaders[3]
centers_data_loaders[3]=tmp

options = {'K': K, 'l_epoch': local_epoch, 'B': batch_size, 'g_epoch': global_epoch, 'clients': clients,
           'l_lr':local_lr, 'g_lr':global_lr, 'dataloader':centers_data_loaders}

In [4]:
scaffold = Scaffold(options)
hist_loss = scaffold.train_server(global_epoch, local_epoch, global_lr, local_lr)
scaffold.test(all_test_loader)

*** global_epoch: 1 ***


/home/madtoinou/documents/venvs/tml/lib/python3.7/site-packages/monai/losses/dice.py:150: UserWarning: single channel prediction, `include_background=False` ignored.
  warnings.warn("single channel prediction, `include_background=False` ignored.")


center1 loss: 0.929850161075592
center2 loss: 0.9844788908958435
center4 loss: 1.0
*** global_epoch: 2 ***
center1 loss: 0.926748514175415
center2 loss: 0.9866360425949097
center4 loss: 1.0
*** global_epoch: 3 ***
center1 loss: 0.9369733929634094
center2 loss: 0.973732590675354
center4 loss: 1.0
*** global_epoch: 4 ***
center1 loss: 0.9189035892486572
center2 loss: 0.9476808309555054
center4 loss: 1.0
*** global_epoch: 5 ***
center1 loss: 0.8083707690238953
center2 loss: 0.9825140833854675
center4 loss: 1.0
*** global_epoch: 6 ***
center1 loss: 0.9217368364334106
center2 loss: 0.9492074847221375
center4 loss: 1.0
*** global_epoch: 7 ***
center1 loss: 0.8111785650253296
center2 loss: 0.9824970364570618
center4 loss: 1.0
*** global_epoch: 8 ***
center1 loss: 0.9217018485069275
center2 loss: 0.9493539333343506
center4 loss: 1.0
*** global_epoch: 9 ***
center1 loss: 0.8116187453269958
center2 loss: 0.9824921488761902
center4 loss: 1.0
*** global_epoch: 10 ***
center1 loss: 0.92167162895202

In [5]:
lt=0
for j in centers_data_loaders:
    lt +=len(j[0])

options.update({'trainloaders_lengths':lt,
                 'beta_val':0.9})
        
options['weighting_scheme']="BETA"
fedbeta = FedAvg(options)
hist_loss = fedbeta.train_server(global_epoch, local_epoch, global_lr, local_lr)
fedbeta.test(all_test_loader)

*** global_epoch: 1 ***
center1 loss: 0.8623128533363342
center2 loss: 0.9571033120155334
center4 loss: 1.0
*** global_epoch: 2 ***
center1 loss: 0.8366109132766724
center2 loss: 0.944184422492981
center4 loss: 1.0
*** global_epoch: 3 ***
center1 loss: 0.7991296052932739
center2 loss: 0.9258918762207031
center4 loss: 1.0
*** global_epoch: 4 ***
center1 loss: 0.7459497451782227
center2 loss: 0.897691011428833
center4 loss: 1.0
*** global_epoch: 5 ***
center1 loss: 0.689969539642334
center2 loss: 0.8663777112960815
center4 loss: 1.0
*** global_epoch: 6 ***
center1 loss: 0.5984511375427246
center2 loss: 0.8303675055503845
center4 loss: 1.0
*** global_epoch: 7 ***
center1 loss: 0.5137276649475098
center2 loss: 0.7994442582130432
center4 loss: 1.0
*** global_epoch: 8 ***
center1 loss: 0.4466356635093689
center2 loss: 0.7846068739891052
center4 loss: 1.0
*** global_epoch: 9 ***
center1 loss: 0.5080825090408325
center2 loss: 0.7607886791229248
center4 loss: 1.0
*** global_epoch: 10 ***
center

In [6]:
options['weighting_scheme']="FEDAVG"
fedavg = FedAvg(options)
hist_loss = fedavg.train_server(global_epoch, local_epoch, global_lr, local_lr)
fedavg.test(all_test_loader)

*** global_epoch: 1 ***
center1 loss: 0.8458237648010254
center2 loss: 0.9466336369514465
center4 loss: 1.0
*** global_epoch: 2 ***
center1 loss: 0.8017458915710449
center2 loss: 0.9237979650497437
center4 loss: 1.0
*** global_epoch: 3 ***
center1 loss: 0.731121838092804
center2 loss: 0.8920515775680542
center4 loss: 1.0
*** global_epoch: 4 ***
center1 loss: 0.6568475961685181
center2 loss: 0.8489447832107544
center4 loss: 1.0
*** global_epoch: 5 ***
center1 loss: 0.55265212059021
center2 loss: 0.8136163353919983
center4 loss: 1.0
*** global_epoch: 6 ***
center1 loss: 0.4395267367362976
center2 loss: 0.7830955982208252
center4 loss: 1.0
*** global_epoch: 7 ***
center1 loss: 0.3592684864997864
center2 loss: 0.7576908469200134
center4 loss: 1.0
*** global_epoch: 8 ***
center1 loss: 0.36083221435546875
center2 loss: 0.7232249975204468
center4 loss: 1.0
*** global_epoch: 9 ***
center1 loss: 0.261584997177124
center2 loss: 0.754959225654602
center4 loss: 1.0
*** global_epoch: 10 ***
center1

- batch_size is limited by the maximum number of slice for one patient, due to the Monai Implementation.
- we should be able to increase it and only be dependent on the number of patient?
- batchsize of 3, with 2 patients with 2 scans : one patient might be picked twice and crash the dataloader

- winner of the contest, using all the modalities 0.56 or 0.55 
- with T-max centralized : 0.4
- with T-max decentralized : 0.26
- play with the learning_rate to see how the evolution of the loss evolve

# Observations about the training
The loss is not stable from one global epoch to the next
## Only 3 sites
- site 2 seems unstable; after 10 global epochs, the loss becomes nan value and this get propagated to the central model


# Misc
C variable(attribute is not used anywhere

# TODO

- clean OO implementation for Scaffold, FedAvg, Softmax and B-weighting
- find a way to increase the batch size; these tiny batches are not satisfying. Too sensible to data augmentation techniques...